activerecord/lib/active_record/core.rb

```diff
@@ -551,12 +551,13 @@ def encode_with(coder)
     # Note also that destroying a record preserves its ID in the model instance, so deleted
     # models are still comparable.
     def ==(comparison_object)
-      return super if new_record?
+      return super if new_record? && !self.id_changed?
```

Comment:
I think this will create undefined behaviour since this is overriding not only `==` but also `eql?`. When implementing `hash`/`eql?` it's necessary that when `a.eql?(b)` it must also be true that `a.hash == b.hash`. Which isn't true here anymore.

Reply:
Good catch. We changed `#hash` too.